contracts/BC_fusion/StakeHub.sol

```diff
@@ -0,0 +1,876 @@
+                        consensusAddress: consensusAddress,
+                        feeAddress: payable(0),
+                        BBCFeeAddress: address(0),
+                        votingPower: votingPowers[i],
+                        jailed: false,
+                        incoming: 0
+                    })
+                );
+                _eligibleValidatorVoteAddrs.push(_validators[operatorAddress].voteAddress);
+            } else {
+                IBSCValidatorSet.Validator storage eVal = _eligibleValidators[j];
+                eVal.consensusAddress = consensusAddress;
+                eVal.votingPower = votingPowers[i];
+                _eligibleValidatorVoteAddrs[j] = _validators[operatorAddress].voteAddress;
+            }
+            ++j;
+        }
+    }
+
+    function downtimeSlash(address consensusAddress, uint256 height) external onlyInitialized onlySlash {
+        address operatorAddress = _consensusToOperator[consensusAddress];
+        require(operatorAddress != address(0), "INVALID_CONSENSUS_ADDRESS"); // should never happen
+        Validator storage valInfo = _validators[operatorAddress];
+        require(valInfo.poolModule != address(0), "VALIDATOR_NOT_EXIST");
+
+        // check slash record
+        bytes32 slashKey = _getSlashKey(operatorAddress, height, SlashType.DownTime);
+        SlashRecord storage record = _slashRecords[slashKey];
+        require(record.slashHeight == 0, "SLASHED");
+
+        // slash
+        uint256 slashAmount = IStakePool(valInfo.poolModule).slash(downtimeSlashAmount);
+        _jailValidator(valInfo);
+
+        // record
+        record.jailUntil = block.timestamp + downtimeJailTime;
+        record.slashAmount = slashAmount;
+        record.slashHeight = uint248(height);
+        record.slashType = SlashType.DownTime;
+
+        if (valInfo.jailUntil < record.jailUntil) {
+            valInfo.jailUntil = record.jailUntil;
+        }
+
+        emit ValidatorSlashed(
+            operatorAddress, record.jailUntil, record.slashAmount, record.slashHeight, SlashType.DownTime
+        );
+
+        _sync(valInfo.poolModule, operatorAddress);
+    }
+
+    function maliciousVoteSlash(bytes calldata _voteAddr, uint256 height) external onlyInitialized onlySlash {
+        address operatorAddress = _voteToOperator[_voteAddr];
+        require(operatorAddress != address(0), "INVALID_CONSENSUS_ADDRESS"); // should never happen
+        Validator storage valInfo = _validators[operatorAddress];
+        require(valInfo.poolModule != address(0), "VALIDATOR_NOT_EXIST");
+
+        // check slash record
+        bytes32 slashKey = _getSlashKey(operatorAddress, height, SlashType.MaliciousVote);
+        SlashRecord storage record = _slashRecords[slashKey];
+        require(record.slashHeight == 0, "SLASHED");
+
+        // slash
+        uint256 slashAmount = IStakePool(valInfo.poolModule).slash(doubleSignSlashAmount);
+        _jailValidator(valInfo);
```

Comment:
`numOfJailed += 1` may be wrong if you jail a already jailed val.

Reply:
Done